### PR TITLE
Migrate media-library tests from enzyme to react testing library

### PR DESCRIPTION
**Pull request by vikneshwar**
#### Changes proposed in this Pull Request

This PR refactors the following test files to use `@testing-library/react` instead of `enzyme`.

- `client/my-sites/media-library/test/data-source.jsx`
- `client/my-sites/media-library/test/list-item-image.jsx`
- `client/my-sites/backup/backup-date-picker/test/index.jsx`

Part of #63409.

#### Testing instructions

Verify tests still pass: `yarn run test-client client/my-sites/media-library/ client/components/jetpack/daily-backup-status/`

### CODE REVIEW

File: client/my-sites/media-library/test/data-source.jsx
Comment (spraveenitpro) [inline client/my-sites/media-library/test/data-source.jsx]: What about using `queryBy` instead of `getBy` methods when asserting 
  with `.toBeInTheDocument` because `getBy` will throw an error of itself if the element isn't found.  `expect( screen.queryByText( 'WordPress.com' ) ).toBeInTheDocument(); instead of expect( screen.getByText( 'WordPress.com' ) ).toBeInTheDocument()`

File: client/my-sites/media-library/test/data-source.jsx
Comment (spraveenitpro) [inline client/my-sites/media-library/test/data-source.jsx]: What about using `queryByText` for both these.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (spraveenitpro) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Just a nitpick but is there a need for a variable `yesterdayTextNode` here? I like to read it being explicit and idiomatic like `expect( screen.queryByText( 'WordPress.com' ) ).toBeInTheDocument();`

File: client/my-sites/media-library/test/data-source.jsx
Comment (vikneshwar) [reply client/my-sites/media-library/test/data-source.jsx]: updated it

File: client/my-sites/media-library/test/data-source.jsx
Comment (vikneshwar) [reply client/my-sites/media-library/test/data-source.jsx]: updated it

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (vikneshwar) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: Storing it in a variable because we do multiple assertions in that test, else we will have to do redundant query calls

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Noting that tests that verify the DOM structure aren't very valuable when that structure is static. The only time when I see such tests as having value is when the `className` actually depends on a prop or any other internal logic.

This applies to multiple assertions in this PR, where the `className` is static and not changeable by any business logic.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: `.toBeInTheDocument()` is generally fine, but for visible elements, I always recommend `.toBeVisible()`, as something can be in the DOM tree but be invisible to the user. And what really matters is whether the user can interact with it. 

This applies to a few more instances where we use `toBeInTheDocument()`.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Why are we not using `user.keyboard()` from `@testing-library/user-event` here? 

This looks like an opportunity to fix the underlying issue, and start relying on the universal `event.code` instead. See https://github.com/WordPress/gutenberg/pull/43437/ for a similar fix.

This applies to the rest of the PR where we use `fireEvent`.

File: client/my-sites/media-library/test/data-source.jsx
Comment (tyxla) [inline client/my-sites/media-library/test/data-source.jsx]: Do you mind using destructuring of arguments to avoid props object access like this? Something like:

```suggestion
	return ( { 'data-source': dataSource, children } ) => <div data-test={ dataSource }>{ children }</div>;
```

File: client/my-sites/media-library/test/data-source.jsx
Comment (tyxla) [inline client/my-sites/media-library/test/data-source.jsx]: It seems like with these changes, we no longer need to pass `data-source` to the `Popover` mock?

File: client/my-sites/media-library/test/data-source.jsx
Comment (tyxla) [inline client/my-sites/media-library/test/data-source.jsx]: The official [recommendation](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-query-variants-for-anything-except-checking-for-non-existence) suggests that we use `query*` screen queries only for checking non-existence (like on the line below).

```suggestion
			expect( screen.getByText( 'Google Photos' ) ).toBeInTheDocument();
```

This applies to at least one more instance in this PR.

File: client/my-sites/media-library/test/list-item-image.jsx
Comment (tyxla) [inline client/my-sites/media-library/test/list-item-image.jsx]: Let's use destructuring here, too:

```suggestion
	return ( { children, src } ) => <div data-testid={ src }>{ children }</div>;
```

**Review by tyxla — COMMENTED** [review]
Thanks a lot for working on this 🙌 It's lovely to see more helping hands for https://github.com/Automattic/wp-calypso/issues/63409!

This generally looks good, I've left a few suggestions, mostly stylistic ones, to improve the quality of the tests before we ship it.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (vikneshwar) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: Hi @tyxla, I decided to do `className` check because I wanted to make sure the date `yesterdayTextNode` appears only under the previous button in the date picker but not in any other place in the document(should not appear in the current date or next date). If I just have to check `toBeInTheDocument` the test might result in a false positive if it appears in any other place. And also there doesn't seem to be any role attached to the previous button in the date picker, so I can't query with `getByRole` to correctly assert that the date shows up only in the location i intended to.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (vikneshwar) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: Hi @tyxlam, I have updated to use `user.type`, but I cannot get `user.keyboard` to work, which I am not exactly sure why. I guess it has to do with the button not being focused when I trigger `user.keyboard`. I did try to focus it with the `button.focus()` before I trigger the keyboard event, but that didn't seem to help. Can you help me understand what I am doing wrong here?

**Comment by vikneshwar** [comment]
Hi @flootr and @tyxla, I updated the PR with the suggestions. I hope it looks good now. Sorry about the delay in addressing it.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: I understand, however, it's often the case that to properly assert that the structure is correct we often need a ton more  assertions. If we really want to do that, that's when snapshot tests come handy. Have you considered adding one in this scenario, rather than doing DOM traversing and checking for classnames?

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: This is generally okay to leave as-is. I've had similar problems that required refactoring components to go from `keyCode` to `code` or `key` instead. 

However, looking at the component code:

https://github.com/Automattic/wp-calypso/blob/9fb7ac149e55361c0c3a9639ad0caa5e15fc88b0/client/my-sites/backup/backup-date-picker/index.tsx#L27

I notice that it uses `key` and compares it to ` ` (space), which I expect will work correctly, because it corresponds to the [mapping](https://github.com/testing-library/user-event/blob/2852509f8e4d5ca28ec494cc85dc4020b0e44b45/src/keyboard/keyMap.ts#L20) of `.keyboard()` `@testing-library/user-event` convenience helper.

Also, another thing to keep in mind is that `@testing-library/user-event` will generally behave as the user would, so if we use [`.type()`](https://testing-library.com/docs/user-event/utility#type), it will auto-focus the element before actually firing `.keyboard()` on it. 

An often occurring mistake is not considering that events are async, and not waiting for them to finish. To fix that, we make the test function `async()` and add `await` before the `user.type()`. call.

If that doesn't help, I'm happy to help diagnose what's wrong if you push the code that fails. Leaving it up to you, because using `@testing-library/user-event` is not a requirement, but more of a recommendation.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: The current way of writing those async tests is not ideal, since the order is a bit confusing - the assertions are at the top, and the rendering is at the bottom. To improve this, I'd suggest to:

* Remove the promise
* Make the test function `async`.
* Add an `await` before any `user.*` event trigger, so the assertions would be run after it fully finishes execution.
* Move the assertions to the bottom of the tests.

This applies to the rest of the same style tests that directly rely on a `Promise`.

What do you think?

File: client/my-sites/media-library/test/list-item-image.jsx
Comment (tyxla) [inline client/my-sites/media-library/test/list-item-image.jsx]: Should we use `toBeVisible` as a more specific version of `toBeInTheDocument()`?

**Review by tyxla — COMMENTED** [review]
Thanks @vikneshwar, this is wonderful work ❤️ 

I've left a few suggestions that I think we could address before merging this.

Let me know what you think and we can take it from there!

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (vikneshwar) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: Hi @tyxla, I have updated the component to use `code` and also updated the test so that we can `await` before the assertion, yet the keyboard event is not getting triggered, I have pushed the changes with what I tried. I hope you can take a look and help me figure out what I am missing. Thanks

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (vikneshwar) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: I have updated to use the async function instead of a promise. Thanks

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Instead of directly accessing `.mock.calls`, let's use a more expressive and idiomatic way:

```suggestion
		expect( dispatchMock ).toHaveBeenCalledWith(
			expect.objectContaining( {
				meta: {
					analytics: [
						expect.objectContaining( {
							payload: expect.objectContaining( {
								name: 'calypso_jetpack_backup_date_previous',
							} ),
						} ),
					],
				},
			} )
		);
```

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Same as above:

```suggestion
		expect( dispatchMock ).toHaveBeenCalledWith(
			expect.objectContaining( {
				meta: {
					analytics: [
						expect.objectContaining( {
							payload: expect.objectContaining( {
								name: 'calypso_jetpack_backup_date_next',
							} ),
						} ),
					],
				},
			} )
		);
```

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: This will be a problem, because calling `moment()` will return the date for today, but on the next day, it will be for tomorrow. This will cause a different snapshot and it will keep failing every day. Should we just use a static date here to address this problem?

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: This will actually select the wrapping `span` and not the button. So when we focus and trigger a keydown event on it, this won't do anything. Instead, to focus the button, we could do something like:

```suggestion
		const button = screen.getAllByRole( 'button' )[ 0 ];
```

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Same as above, but grabbing the second found button.

```suggestion
		const button = screen.getAllByRole( 'button' )[ 1 ];
```

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: Thanks, I'm going to leave a separate comment in my review.

File: client/my-sites/backup/backup-date-picker/index.tsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/index.tsx]: This is nice 👍 Let's just make sure that nothing is relying on the old behavior anymore.

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (tyxla) [inline client/my-sites/backup/backup-date-picker/test/index.jsx]: Just a side question: did we mean this refactor to be in a separate PR? Or do you feel it's somehow related to the media library tests?

**Comment by tyxla** [comment]
@vikneshwar thanks for the great work here! I've left some feedback and suggestions.

This looks pretty close already and I think it should be good to go once feedback is addressed 🚀

File: client/my-sites/backup/backup-date-picker/test/index.jsx
Comment (vikneshwar) [reply client/my-sites/backup/backup-date-picker/test/index.jsx]: Agree. I have moved the changes to #68933

**Review by tyxla — APPROVED** [review]
This looks awesome - nice work! Let's 🚀

**Pull request merged**
